test/test-parse-warning.cpp

```diff
@@ -25,655 +25,770 @@ using ::testing::IsEmpty;
+              }));
+        }
+      }
     }
-  }
-}
 
-TEST_F(test_parse_warning,
-       warn_on_pointless_compare_against_literals_complex_expressions) {
-  {
-    test_parser p(
-        u8"({} == {} && (x) === [1, 2, 3]) || ((/pattern/) == y.prop)"_sv,
-        capture_diags);
-    p.parse_and_visit_expression();
-    EXPECT_THAT(
-        p.errors,
-        ElementsAreArray({
-            DIAG_TYPE_OFFSETS(p.code,
-                              diag_pointless_comp_against_object_literal,
-                              equals_operator, strlen(u8"({} "), u8"=="_sv),
-            DIAG_TYPE_OFFSETS(
-                p.code, diag_pointless_strict_comp_against_array_literal,
-                equals_operator, strlen(u8"({} == {} && (x) "), u8"==="_sv),
-            DIAG_TYPE_OFFSETS(
-                p.code, diag_pointless_comp_against_regular_expression_literal,
-                equals_operator,
-                strlen(u8"({} == {} && (x) === [1, 2, 3]) || ((/pattern/) "),
-                u8"=="_sv),
-        }));
-  }
-  {
-    test_parser p(u8"x === y || ({}) != obj.prop"_sv, capture_diags);
-    p.parse_and_visit_expression();
-    EXPECT_THAT(
-        p.errors,
-        ElementsAreArray({
-            DIAG_TYPE_OFFSETS(
-                p.code, diag_pointless_comp_against_object_literal,
-                equals_operator, strlen(u8"x === y || ({}) "), u8"!="_sv),
-        }));
+    TEST_F(test_parse_warning,
+           warn_on_pointless_compare_against_literals_complex_expressions)
+    {
+      {
+        test_parser p(
+            u8"({} == {} && (x) === [1, 2, 3]) || ((/pattern/) == y.prop)"_sv,
+            capture_diags);
+        p.parse_and_visit_expression();
+        EXPECT_THAT(
+            p.errors,
+            ElementsAreArray({
+                DIAG_TYPE_OFFSETS(p.code,
+                                  diag_pointless_comp_against_object_literal,
+                                  equals_operator, strlen(u8"({} "), u8"=="_sv),
+                DIAG_TYPE_OFFSETS(
+                    p.code, diag_pointless_strict_comp_against_array_literal,
+                    equals_operator, strlen(u8"({} == {} && (x) "), u8"==="_sv),
+                DIAG_TYPE_OFFSETS(
+                    p.code, diag_pointless_comp_against_regular_expression_literal,
+                    equals_operator,
+                    strlen(u8"({} == {} && (x) === [1, 2, 3]) || ((/pattern/) "),
+                    u8"=="_sv),
+            }));
+      }
+      {
+        test_parser p(u8"x === y || ({}) != obj.prop"_sv, capture_diags);
+        p.parse_and_visit_expression();
+        EXPECT_THAT(
+            p.errors,
+            ElementsAreArray({
+                DIAG_TYPE_OFFSETS(
+                    p.code, diag_pointless_comp_against_object_literal,
+                    equals_operator, strlen(u8"x === y || ({}) "), u8"!="_sv),
+            }));
+      }
+
+      TEST_F(test_parse_warning, warn_on_mistyped_strict_inequality_operator)
```

Comment:
Issue 1: You need to run `./tools/update-translator-sources` again.

Issue 2: You added this `TEST_F` in the middle of an existing `TEST_F`. Put it outside (e.g. move the `}` on line 788 to above line 705).

Reply:
Thank you, that worked.
Next question, I am trying to find the part in the code that you described as "In parser::parse_expression_remainder, when we create an expression::non_null_assertion." could you help me locate this part in the code? Is it around line 207?

Reply:
> Is it around line 207?

No. Creation of a expression::non_null_assertion is on line 1783 of parse-expression.cpp:
https://github.com/quick-lint/quick-lint-js/blob/b9013cd85bfe3c7fd2b754d76a1aa9cf68aec1f6/src/quick-lint-js/fe/parse-expression.cpp#L1782-L1784

Reply:
Ah okay. So, there I would check for the left-hand side is not in parenthesis, and that to the right of the "!" is "==", right? Then we would call the warning.

Reply:
try it out